aaronlademann-wf/react_material_ui
# CPLAT-11503 CSS in JS

### MUI CSS in JS Investigation

#### Changes
- Add `makeStyles`, `styled`, `withStyles` APIs
- Add style_utils.dart to for utilities and shared classes
- Add a `ButtonClasses` map interface to facilitate overriding the `classes` prop
- Add examples for all of the above

#### QA'ing
To qa:
- Pull down the changes and serve the `example` dir (both with dartdevc and dart2js)
- The three style APIs have almost identical example pages to demonstrate how they behave. 
  - Inspect the DOM and make sure the generated style tags and classes are expected.
- For the "misc" examples (right now really just the "classes" prop examples), also just inspect the DOM and make sure the classes appended as expected

#### Regarding statically typing the "classes" prop on MUI components:
The AC mentioned exploring this, so here is the outcome of the exploration:

##### Summary
Overall, it's probably worth doing. Adding a statically typed interface could definitely help reduce typos, which could be annoying to find being a string in the middle of a `Map` of strings. It also seems like the type of job that could easily be automated. If generating the interfaces _couldn't_ be automated, creating `MapView`s like the example here is really quick and easy, the maintenance would be the only issue. 

That being said, it'd probably be a low priority because adding an interface to facilitate connecting a generated class name to a pre-existing one doesn't save a ton of trouble. Additionally, I'm unsure how often consumers would use the `classes` prop. Perhaps a lot (we'll definitely use it a lot at the platform level), but depending on how much we wrap things ourselves or provide theming, it may not be necessary to make it the easiest API to use if it takes too much effort.

##### What value is gained from typing them?
- Autocompletion
  - This feedback would likely lead to less typos / bugs
- Fluent style value setting. This takes away the burden of looking at lots of strings in the map, which seems preferable

##### What non-value is gained from typing them?
- Lots of code & potential maintenance 
- It doesn't really help _much_. The typing is `Map<String, String>`, so it's not clarifying any typing questions at the property level.
